Websocket console support

jukkar opened the pull request:
Initial support for websocket based console. A helper script for testing can be found in net-tools https://github.com/zephyrproject-rtos/net-tools/pull/73
This PR was created in order to test and experiment with http-server/websocket support.

jukkar committed:
shell: backend: Create a websocket transport backend

This creates a websocket based shell backend that is used to
implement a websocket console that can be connected using a browser.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: echo-server: Add websocket console support

Add enablers that can be used to setup websocket console
for this sample.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
Just FYI at this point cc: @rlubos, @cfriedt

rlubos reviewed (commented):
Looks good overall, but wouldn't it make sense to have a separate sample sample for this feature?

jukkar commented:
> wouldn't it make sense to have a separate sample sample for this feature?

I was thinking about that, and in 1.13 we had a separate app for it. The issue with it is that the app then does nothing useful because there is nothing "else" in the app. That is why I placed the support in one of our already existing app. I considered HTTP server sample app, but then did it for echo-server so that we can test the code even if there is no other HTTP server users in the application. I had hoped we could make this console support a library so that user could enable it by a Kconfig option but that is not easy because of TLS support which needs to be enabled by application.

jukkar commented:
This still needs TLS support enabled and tested, I will do it next.
If you want to test the current functionality you can do this:

1. Enable zeth, by executing `cd tools/net-tools; ./net-setup.sh`
2. Compile the echo-server sample
```
west build -p -b native_sim -d ~/zephyrproject/build/ws-console ~/zephyrproject/zephyr/samples/net/sockets/echo_server/ -- \
-DCONFIG_NATIVE_UART_0_ON_STDINOUT=n \
-DCONFIG_LOG_BACKEND_UART=y \
-DCONFIG_SHELL_LOG_BACKEND=n \
-DCONFIG_NATIVE_UART_AUTOATTACH_DEFAULT_CMD="\"gnome-terminal -- screen %s\"" \
-DCONFIG_OBJ_CORE=y -DCONFIG_NET_SOCKETS_OBJ_CORE=y \
-DCONFIG_NET_SHELL=y -DCONFIG_NET_LOG=y \
-DCONFIG_NET_SOCKETS_LOG_LEVEL_DBG=n \
-DCONFIG_NET_HTTP_SERVER_LOG_LEVEL_DBG=n \
-DCONFIG_NET_WEBSOCKET_LOG_LEVEL_DBG=n \
-DCONFIG_SHELL_WEBSOCKET_INIT_LOG_LEVEL_DBG=n \
-DOVERLAY_CONFIG="overlay-ws-console.conf" \
-DCONFIG_SHELL_WEBSOCKET_BACKEND_COUNT=2

if [ $? -eq 0 ]; then
   ~/zephyrproject/build/ws-console/zephyr/zephyr.exe -attach_uart
fi
```
And then point your browser to `http://192.0.2.1/console`

jukkar commented:
* Added actual console support that uses the websocket shell transport for console output. I was not really able to test this. I am seeing that the websocket console driver init is called but I am not seeing the output function to be called. Any help what options to set/unset in Kconfig would be appreciated.

cfriedt commented:
I'll take a look today.

cfriedt commented on drivers/console/websocket_console.c:
```diff
@@ -0,0 +1,126 @@
+ */
+
+
+#include <zephyr/kernel.h>
```
Since the zephyr console is a singleton now, I think we might actually need to have DT bindings for this, if we want it to look and feel like a console and get all of the right hooks. But that implies that it would also need to be active immediately on init.

A slightly alternative approach would be to dynamically set up and tear down a zephyr shell per websocket.

Currently, it still seems like the main UART device gets the console / shell.

I could be totally wrong about that though, since I still have not completely gotten the web page to load. Likely I need some iptables rules to forward a port from the physical interface to the tap interface on my build machine.

jukkar replied:
Ok, thanks for the info. I wonder should I remove the actual console support if it is not working as expected and just leave the shell support, wdyt?

cfriedt replied:
It might be as easy as defining a Devicetree binding, actually, so could be an easy win.

I enabled the RTT console today, which (in theory) is not very different from the ws console, and that was about as easy to enable.

jukkar replied:
> It might be as easy as defining a Devicetree binding, actually, so could be an easy win.

Ok, I will investigate this.

psztoch commented:
How can I resolve console access authorization?

jukkar commented:
> Console access should be protected from unauthorized access.
> How do you propose to solve this?

That is easy to answer, as this is just a demo I do not plan to provide anything for this :see_no_evil: 
Currently one needs to create an application for this with http-server enabled. For access control, something would be needed to be created in the http server side.
Lets figure out how to solve this if I manage to finish this.

psztoch commented:
> > Console access should be protected from unauthorized access.
> > How do you propose to solve this?
> That is easy to answer, as this is just a demo I do not plan to provide anything for this 🙈 Currently one needs to create an application for this with http-server enabled.
A complete demo will definitely be of greater value to everyone. :-)

> For access control, something would be needed to be created in the http server side. Lets figure out how to solve this if I manage to finish this.
How about basing authorization on a password that is stored in the settings subsystem?
Basic authorization in HTTP should be enough.

cfriedt commented:
I'm not entirely convinced that password-based security is needed (or very useful) here. There are certainly many other samples in Zephyr that enable console access with zero "protection" amd even less of a hurdle to jump through (e.g. a USB port is sufficient).

My 2¢ - we should (eventually) have a more fully featured sample for the web server where an SSL key is generated after the first power-on, and the first user to visit a page is required to create the admin password.

That also begs the question, about how / where the password is stored. Is it encrypted? Is the encryption key stored in a secure location, etc.

Personally, I think there is room for simple samples (used to convey a key subset of concepts) and more "production ready" samples with security in mind.

kartben commented:
> Personally, I think there is room for simple samples (used to convey a key subset of concepts) and more \"production ready\" samples with security in mind.

Yes, this :)
Related: https://github.com/zephyrproject-rtos/zephyr/issues/67021

kartben commented on include/zephyr/net/http/hpack.h:
```diff
@@ -94,6 +94,12 @@ enum http_hpack_static_key {
+#if defined(CONFIG_HTTP_SERVER)
+#define HTTP_SERVER_HUFFMAN_DECODE_BUFFER_SIZE CONFIG_HTTP_SERVER_HUFFMAN_DECODE_BUFFER_SIZE
+#else
+#define HTTP_SERVER_HUFFMAN_DECODE_BUFFER_SIZE 0
```
Do we allow zero-length arrays in Zephyr? Would be better to avoid it I think?
https://github.com/zephyrproject-rtos/zephyr/issues/63845

jukkar replied:
This is only used to compile test the struct where the array is used.

kartben commented on samples/net/sockets/echo_server/src/common.h:
```diff
@@ -82,6 +82,15 @@ static inline int init_vlan(void)
+{
+	return 0;
+}
+#endif /* CONFIG_NET_VLAN */
```
Copy paste error

kartben reviewed (commented):
Disclaimer: I gave ChatGPT's new o1 model a try :)

kartben reviewed (dismissed):
"commented" instead of "requesting changes" earlier

jukkar committed:
logging: backend: Add websocket backend

Allow logging output to websocket console. Requires that
websocket shell backend and websocket HTTP server support
is also enabled.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Changed according to comments
* Rebased to latest upstream

This now supports also logging output. So I ditched the console driver and implemented websocket logging backend. Works fine now. The "console" output is also seen in websocket console together with shell output.

To try this with native-sim, you can do this:
```
west build -p -b native_sim -d ../build/ws-console samples/net/sockets/echo_server/ -- \
-DCONFIG_NATIVE_UART_AUTOATTACH_DEFAULT_CMD="\"gnome-terminal -- screen %s\""  \
-DCONFIG_NET_SHELL=y -DCONFIG_NET_LOG=y -DOVERLAY_CONFIG="overlay-ws-console.conf"
```
and then connect to the console using your host browser [http://192.0.2.1](http://192.0.2.1)

cfriedt commented on samples/net/sockets/echo_server/overlay-ws-console.conf:
```diff
@@ -0,0 +1,13 @@
+CONFIG_EVENTFD=y
+CONFIG_POSIX_API=y
+CONFIG_FDTABLE=y
+CONFIG_POSIX_API=y
```
POSIX API appears twice. I feel like there is some kind of Dos Equis humour to add here.

> I don't always use the websocket console, but when I do..

😅

cfriedt replied:
How did that meme become 18 years old all of a sudden 😬?

jukkar replied:
Updated the prj.conf. Dunno why I did that, but I suppose it is always better to have things mentioned twice, just to be in safe side :grin:

cfriedt reviewed (commented):
Just a minor nit in prj.conf, otherwise looks great to me.

I'm ready to approve, but could give it a spin if needed.

jukkar commented:
> I'm ready to approve, but could give it a spin if needed.

It would indeed be great if someone else tries this too. The web page is ugly as hell, but hey, I am not a web designer so this is what I could do :sunglasses:  Patches are certainly welcome to make the console page look nice.

cfriedt commented:
Is this just about mergeable?

Maybe we should just get it in - the html / CSS can always be beautified later.

jukkar commented:
This should be ready to go in. The web page is usable but would need some TLC in the future. I will rebase this against latest upstream to make sure this still compiles ok.
